lib/spack/spack/mixins.py

```diff
@@ -0,0 +1,177 @@
+        if PackageMixinsMeta._methods_to_be_added:
+            attr_dict.update(PackageMixinsMeta._methods_to_be_added)
+            PackageMixinsMeta._methods_to_be_added.clear()
+
+        attr_fmt = '_InstallPhase_{0}'
+
+        # Copy the phases that needs it to the most derived classes
+        # in order not to interfere with other packages in the hierarchy
+        phases_to_be_copied = list(
+            PackageMixinsMeta._add_method_before.keys()
+        )
+        phases_to_be_copied += list(
+            PackageMixinsMeta._add_method_after.keys()
+        )
+
+        for phase in phases_to_be_copied:
+
+            attr_name = attr_fmt.format(phase)
+
+            # Here we want to get the attribute directly from the class (not
+            # from the instance), so that we can modify it and add the mixin
+            # method to the pipeline.
+            phase = getattr(cls, attr_name)
+
+            # Due to MRO, we may have taken a method from a parent class
+            # and modifying it may influence other packages in unwanted
+            # manners. Solve the problem by copying the phase into the most
+            # derived class.
+            setattr(cls, attr_name, phase.copy())
+
+        # Insert the methods in the appropriate position
+        # in the installation pipeline.
+
+        for phase in PackageMixinsMeta._add_method_before:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_before.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_before.clear()
+
+        for phase in PackageMixinsMeta._add_method_after:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_after.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_after.clear()
+
+        super(PackageMixinsMeta, cls).__init__(name, bases, attr_dict)
+
+
+def filter_compiler_wrappers(*files, **kwargs):
+    """Substitutes any path referring to a Spack compiler wrapper with the
+    path of the underlying compiler that has been used.
+
+    If this isn't done, the files will have CC, CXX, F77, and FC set to
+    Spack's generic cc, c++, f77, and f90. We want them to be bound to
+    whatever compiler they were built with.
+
+    Args:
+        *files: files to be filtered
+        **kwargs: at present supports the keyword 'after' to specify after
+            which phase the files should be filtered (defaults to 'install').
+    """
+    after = kwargs.get('after', 'install')
+
+    def _filter_compiler_wrappers_impl(self):
+
+        tty.debug('Filtering compiler wrappers: {0}'.format(files))
```

Comment:
I'll remove the debug statement from here, and put `tty.debug` print in `llnl.util.filesystem.filter_file`.